### DISCUSSION TITLE
upgrade kubernetes version

### DISCUSSION THREAD

**Discussion by HendrikZe** [syself/cluster-api-provider-hetzner#656] [Q&A]
Hi, 

what's the best way to upgrade the kubernetes version of an existing cluster?

Thanks in advance

**Comment by privatecoder** [comment]
I would upgrade _regular_ kubeadm-Clusters like [this](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/) however the K8s version is referenced in multiple places, namely

`KubeadmControlPlane`
`KubeadmConfigTemplate md-0`
`KubeadmConfigTemplate md-1`
`MachineDeployment md-0`
`MachineDeployment md-1`

I guess one has to touch those..

**Reply by privatecoder** [reply]
I just tried by editing a `KubeadmConfigTemplate` followed by the corresponding `MachineDeployment` which triggered the creation of a new HCloud machine, cordoning of an existing node followed by a removal.

The new node was running the new K8s version afterwards, however I am not sure how this would work with baremetal (do we need to re-initialize the bm-node?).

**Reply by privatecoder** [reply]
Found this issue here https://github.com/syself/cluster-api-provider-hetzner/issues/591 and also this blog article https://blogs.oracle.com/cloud-infrastructure/post/cluster-api-upgrade-kubernetes-cluster-in-oci

**Comment by batistein** [comment]
In general: 
1. New node images (via packer or via pre/postkubeadmCommands)
2. Create new ProviderMachineTemplates & KubeadmConfigTemplates
3. Change the KubeadmControlPlane configuration more information: https://cluster-api.sigs.k8s.io/tasks/upgrading-clusters.html#how-to-upgrade-the-kubernetes-control-plane-version
4. Change the MachineDeployment
